### PR TITLE
Copied over ssz.DeepEqual as sszutil.DeepEqual from go-ssz

### PR DESCRIPTION
**Pull request by pinglamb**
**What type of PR is this?**

> Other

**What does this PR do? Why is it needed?**

Refs #7845. Copied `ssz.DeepEqual` from `go-ssz`.

https://github.com/prysmaticlabs/prysm/issues/7845#issuecomment-759120206

**Which issues(s) does this PR fix?**

Refs #7845 

**Other notes for review**

There is also another full PR for removing `go-ssz` usage: https://github.com/prysmaticlabs/prysm/pull/8257

### CODE REVIEW

**Comment by pinglamb** [comment]
DeepSource complains the use of `unsafe` which seems to be necessary. Should it be suppressed or there is other way to handle it?

**Comment by prestonvanloon** [comment]
> DeepSource complains the use of `unsafe` which seems to be necessary. Should it be suppressed or there is other way to handle it?

It can be suppressed and I have suppressed it for this file `shared/sszutil/deep_equal.go`.

File: shared/sszutil/deep_equal.go [hunk 1]
```diff
@@ -0,0 +1,193 @@
+package sszutil
```
Comment (prestonvanloon) [inline shared/sszutil/deep_equal.go:1]: I realize there were no tests for this file specifically, but could you add a few sanity tests? Specifically, comparing protobuf generate structs with mixed byte slice values (nil vs empty slice are treated the same).

Comment (0xKiwi) [reply shared/sszutil/deep_equal.go:1]: Same, some tests would be great!

File: shared/sszutil/deep_equal.go [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+// recursive types.
+func deepValueEqual(v1, v2 reflect.Value, visited map[visit]bool, depth int) bool {
+	if !v1.IsValid() || !v2.IsValid() {
+		return v1.IsValid() == v2.IsValid()
+	}
+	if v1.Type() != v2.Type() {
+		return false
+	}
+
+	// if depth > 10 { panic("deepValueEqual") }	// for debugging
```
Comment (terencechain) [inline shared/sszutil/deep_equal.go:37]: Can we remove commented code?

Comment (prestonvanloon) [inline shared/sszutil/deep_equal.go:37]:
```suggestion
```

**Review by prestonvanloon — APPROVED** [review]

File: shared/sszutil/deep_equal.go (hunk 1 above)
Comment (pinglamb) [reply shared/sszutil/deep_equal.go:1]: I have added some test cases, especially for covering the `nil` vs `empty slice`. Feel free to comment.

While I am doing that for `struct`, I found that the `DeepEqual` will panic if the struct doesn't have any public fields, is that an issue? 

```
panic: reflect.Value.Interface: cannot return value obtained from unexported field or method [recovered]
	panic: reflect.Value.Interface: cannot return value obtained from unexported field or method
```
